### DISCUSSION TITLE
Home Assistant setup for negative feed in over analysis

### DISCUSSION THREAD

**Discussion by brodykenrick** [amberelectric/public-api#137] [Show and tell]
I'm a new Amber Customer and my powerwall2 battery has been stuck in VPP-limbo for three weeks (can't get ported properly) so I feel rather exposed to negative feed-in-tariffs at the moment with not much real control of my battery (and not good control to feed in during the evening *). I am in awe of anyone going to Amber without a battery - that's living on the edge!

I made a dashboard with a few charts to be able to see the impact of negative feed in rates on today and tomorrow and plan accordingly (and also just watch things on my phone too much)

When I get my battery over to Amber I'll make a version two and be able to add in some actions (or calls to action for manual operations in the Amber App) and share configs for those interested.

It looks like this.

![Screenshot_20230415-134440](https://user-images.githubusercontent.com/5006130/232181703-33e45a4b-4dd4-44dd-be8f-5623bd9bcc6b.png)

![Screenshot_20230415-142923](https://user-images.githubusercontent.com/5006130/232182926-f0dbfb05-5d69-4028-a44b-2307455784ab.png)

Setup/components/integrations:
* Enphase microinverters
* Tesla Powerwall2
* EV w/ Smappee Charger
* Solcast for solar forecast
* Amber
* Apex Charts

Chart 1:
Battery level
Forecast
Actual Solar
Grid feed in
Times with negative feed in
The potential cost of that negative feed in (sum product of half-hourly forecast feed in rate and forecast solar energy)

Chart 2:
Amber feed in - historical and future (negative accentuate)
Instantaneous total grid cost of my site 

Chart 3:
Today and tomorrow energy forecast
Negative feed in times
Negative feed in sum of energy forecast to be produced

Chart 4:
Amber feed in forecast (only)

Chart 5:
Regular HA energy dashboard card

Table 6:
Solar Forecast for the week - by the numbers

Chart 7:
Emissions displaced/avoided (measured in grams CO2eq displaced from the grid in an averaged five minute interval)

*Note - I have worked out how to force a discharge of the powerwall via time based control with some finicky setup. That's what I did in the morning to test enquiring the battery. You can see that on chart 1.

What I'd really love is:

* The Amber API forecasts going out a few more days in to the future (7 days) 
** then I could chart and project particularly good EV charging days and to avoid feed in costs (our house is trying to be close to absolute zero emissions so it is a juggle)
* The Amber API to have controls for charge/discharge/hold
* The enphase integration to support grid profile change (to simplify export limits in summer with long negative feed in periods with high production)

Lemme know your thoughts or questions or improvement advice (and I'll share the code in a few weeks after I actually am on SmartShift).

**Comment by purcell-lab** [comment]
Great to have you here, and excellent you are using HA.

A couple of comments.

### Negative Feed In

Agreed it is a real pity you cannot get your enphase to switch to zero export.  However all is not lost, if you are on single phase you can take the powerwall off-grid using the core powerwall integration https://www.home-assistant.io/integrations/powerwall.

For single phase off-grid mode stop export and continue to recharge your powerwall from excess solar and run your household.  It will also curtail solar production to the maximum of your household load + battery charging. Note however you won't be able to draw more than the powerwall current limit. I find negative FIT an ideal time to draw additional from the grid as the prices are extermely_low, but you need to have household load available (EV, Pool Heater, AC, ...) 

![Screenshot 2023-04-15 16 40 30](https://user-images.githubusercontent.com/79175134/232192351-d9e9ae7a-af79-4981-8515-a243e1639195.png)

Here is the automation I am using with SolarEdge to switch to zero export when FIT <0 and zero production when the general price <0, you should be able to change the SolarEdge service calls with the powerwall off-grid mode.

![image](https://user-images.githubusercontent.com/79175134/232193954-08d77d8d-7000-41ba-b6a0-84f809245c8d.png)

### Manual Grid Export From Powerwall

You can force powerwall to export to the grid using a Time-Based Control mode hack without SmartShift. If you setup sufficient differentiation with your Utility Rate Plan and buy/sell prices in the Tesla App you can get the powerwall to export to the grid in peak and charge from the grid during off-peak. It is clunky and looks like Tesla have a 15 minute cronjob that updates any setting changes you make in the app to the Utility Rate Plan, so you need to plan ahead.  I have a standing rule & HA automation to export to the grid for an hour at 0600 & 1800 each day.

Set this in the Tesla App:

![Screenshot_20230415-165144](https://user-images.githubusercontent.com/79175134/232193621-54a38bd6-caac-4622-9d36-0c8d4e847d13.png)

Manually switch to Time-Based Control between 0600-0700 & 1800-1900 to see if it works for you, make sure you switch back to Self-Consumption mode or it might start charging the battery from the grid when you don't want.

If this works for you them here is my Home Assistant Automation:

![image](https://user-images.githubusercontent.com/79175134/232194026-fb980395-b6f1-49d5-8a05-15c284b46464.png)
![image](https://user-images.githubusercontent.com/79175134/232193875-23f20452-c236-48e2-ad56-4e3af6349665.png)

### Energy Management for Home Assistant (EMHASS) Add On

If you are up for it - advanced HA, it is a bit complex to set up, have a look at the EMHASS Add On for Home Assistant.  It provides the same optimisation as SmartShift for your battery, but can also optimise and operate your other loads around your house.  Which means it can outperform SmartShift as it can take action with your EV charging, AC or other deferrable loads depending on price inputs. I actually run this in parallel with SmartShift, so either can initiate export or import for my battery.

https://emhass.readthedocs.io/en/latest/intro.html#what-is-this

Here is my EMHASS optimisation for the next 24 hours:
![image](https://user-images.githubusercontent.com/79175134/232194301-0447aa01-cc2e-4449-a4e5-9bd93c1fe70c.png)

A couple of things to note from my EMHASS optimisation:
. Battery (15 kW blue shaded line) discharge to the grid during evening peak until about 7pm
. HVAC (4 kW red line) starts running at reduced power after solar production starts at 0730am
. Pool Heat Pump (5 kWh green line) starts at 9am corresponding with the drop of feed-in to 3c / kWh
. Grid charging of battery (15 kW green shade) starts at 11am corresponding to 4c general price
. Everything switched off by 4pm as prices start to rise and solar production reduces
. Battery (15 kW blue shaded line) discharge to the grid during evening peak until about 7pm
. At this stage it looks like battery is only charging to 50% tomorrow, but that is because it can only see half the evening peak prices, once the additional prices come in, it will schedule full charge to 100% to maximise profit for full discharge tomorrow night as well.

By comparison here is my SmartShift plan for the same period (albeit only upto 0700)
![Screenshot_20230415-171002](https://user-images.githubusercontent.com/79175134/232194464-e572af15-d834-485a-99e3-592a7dd8700f.png)

Happy to assist, if you want to head down this path.

**Reply by brodykenrick** [reply]
Thanks @purcell-lab 

> Great to have you here, and excellent you are using HA.

Thanks for the welcome - I've been using HA for about a year now (after a brief try and give up a few years back). It is working nicely for the amount of consumption monitoring and automation/convenience and is pretty smooth so far.

> Agreed it is a real pity you cannot get your enphase to switch to zero export. However all is not lost, if you are on single phase you can take the powerwall off-grid using the core powerwall integration https://www.home-assistant.io/integrations/powerwall.

Unfortunately I am three phase, so that powerwall path isn't available (unless I buy another two powerwalls).

I know I can get the enphase system to change exports by modifying the grid profiles to change power export limit via the installer API - so I will probably be able to work on that with my installer at least that can be done manually a couple of times a year (it is capped to three grid profiles though - so i'll need to pick something generally good for the season; like no limit=16kW from April to September but capped at 11kW for shoulder and 7kW in the high summer months). Hopefully, can get more dynamic than that though. I'll need to sort that out by December as I don't reckon 85+ kWh export days are in my future anymore.

> You can force powerwall to export to the grid using a Time-Based Control mode hack 

I had been doing something similar today to export but it was much more finnicky than your setup and I didn't realise there was that other custom integration for Tesla that gave you reserve control and the operation modes so it was going to be manual. Thanks @purcell-lab ! I have just installed that integration and I have setup the automations for tomorrow. That will be a nice little automation.

> If you are up for it - advanced HA, it is a bit complex to set up, have a look at the EMHASS Add On for Home Assistant.

I am up for it - but it is down the list a ways. I need to get my head around all of this new Amber setup and operations and then the next goal is to set up a few inkplate dashboards around the house (which has me customising another addon to support some features I want). After that I'll be back on the overall optimisation for the house - but for the next month or so I just want to set up some somewhat set and forget configurations that keep our absolute emissions down and also give us lots of displacement "credits" that we are tracking to make sure we remain significanyly co2eq net negative.

**Comment by madpilot** [comment]
Hi! Welcome!

> What I'd really love is:
>
> * The Amber API forecasts going out a few more days in to the future (7 days)
> ** then I could chart and project particularly good EV charging days and to avoid feed in costs (our house is trying to be close to absolute zero emissions so it is a juggle)

Interesting. It's not possible with the current forecasts, but let me talk to the data team to see what _is_ possible.

> * The Amber API to have controls for charge/discharge/hold

Preliminary work has started. _Hopefully_ by the end of the year. (Cool your heals everyone—preliminary means some architectural charges that are required—there is nothing public that is even close to discussion, let alone testing yet…)

**Reply by brodykenrick** [reply]
Thanks for the welcome, response and the amazing community engaging/creating efforts @madpilot .

> Interesting. It's not possible without current forecasts, but let me talk to the data team to see what is possible.

That would be the kind of thing that I'd dashboard up for the family to see "in 3 days it is a good charging day - but the rest of the week is not so good" or to see something like "there are 5 days on the trot with serious negative feed in risk - so let's use the car as a sink but not charge it all on the first day".

I would expect that others on the general usage side would be keen for extra visibility for deciding when their loads do/don't make sense.

> Preliminary work has started. Hopefully by the end of the year.

Well that is exciting news. I think with @purcell-lab 's help above I can finally stop referencing the Tesla app. When those APIs do get reelased I'll pretty much only ever open HA.

**Reply by purcell-lab** [reply]
> That would be the kind of thing that I'd dashboard up for the family to see "in 3 days it is a good charging day - but the rest of the week is not so good" or to see something like "there are 5 days on the trot with serious negative feed in risk - so let's use the car as a sink but not charge it all on the first day".

I find the SolCast 5 day lookahead is useful to forecast medium term solar production which is aligned with Amber pricing.

![Screenshot_20230416-073313](https://user-images.githubusercontent.com/79175134/232254116-2688f645-8da4-47e8-ad32-5d40b856db8c.png)
 
I like but old fashioned but so does the standard weather report.

![Screenshot_20230416-073725](https://user-images.githubusercontent.com/79175134/232254157-a0c6ae0c-edd6-4391-befa-db63403e6b5d.png)

In terms of getting the family onboard @madpilot blueprint had been great, I have an RYG light in the kitchen and the family check it before they turn on anything big.

https://community.home-assistant.io/t/amber-electric-change-the-colour-of-a-light-based-on-the-current-electricity-price/344612?u=markpurcell

I also have neat little text to speech automations that do a household broadcast when the price hits extremely low or high.

![Screenshot_20230416-074341](https://user-images.githubusercontent.com/79175134/232254355-20526a8f-f720-4f7e-bb26-df88a8f36bbb.png)
![Screenshot_20230416-074407](https://user-images.githubusercontent.com/79175134/232254358-6860a7cd-ecd1-4152-b2fc-d7e9da68ca9e.png)

**Comment by brodykenrick** [comment]
I've been working on the curtailment of enphase generation during periods of significant negative pricing recently.

After the debacle of the Enphase v7 firmware has now passed there is a cool install/DIY permission home assistant custom_component that allows for setting of the power export limits in dynamic response to the amber feed in values. It does this by setting "grid profiles" with a specific value in them for the "power export limit" which causes generation to remain under house/load consumption adn whatever value that limit is set to.

https://github.com/vincentwolsink/home_assistant_enphase_envoy_installer/tree/0.3.5-beta2


There is also a revision for @zimbix awesome standalone ruby app that will allow the same for non-HA users -- https://github.com/ZimbiX/amber-enphase-zero-export-switcher-tool/issues/2



Now, by connecting up with the amber integration/custom component and the enphase installer integration I've been able to setup my system to now curtail/limit exports if negative prices go less than 2c (I'm happy to help the grid with some extra green power at my expense but exporting 15kW at -0.10 for a couple of hours every sunny day through summer is not something I am keen on).

Happy notch days are here!

![Screenshot_20231013-192413-1](https://github.com/amberelectric/public-api/assets/5006130/b78e92dc-fc7a-4db7-be9d-7a064f2d976e)


...Not really, as I'd much prefer to not be throwing away green power though so the more storage we get on the grid the  better.
